File: examples/python/basic.py
"""
Solve PerimeterX / HUMAN Security with Capzy — minimal Python example, `requests` only.

Cost:   from $0.001 per solve (flat)
Speed:  ~10 seconds median

PerimeterX clearance cookies (_px3 / _px2 / _pxhd) are IP-bound — your
proxy is required, and there is no ProxyLess variant. Use the SAME
sticky proxy your downstream HTTP client will replay the cookies on.

Run with:
    pip install requests
    export CAPZY_KEY="capzy_xxxxxxxxxxxxxxxxxxxxxxxx"
    export PROXY_HOST="gw.your-provider.com"
    export PROXY_PORT="10000"
    export PROXY_USER="your-user"
    export PROXY_PASS="your-pass"
    python basic.py
"""

import os
import time

import requests

API_BASE = "https://api.capzy.ai"

# Grab a key for free at https://capzy.ai/auth/register ($0.10 starter credit).
CAPZY_KEY = os.environ["CAPZY_KEY"]

# Your sticky residential / mobile / static-ISP proxy. Datacenter IPs
# will fail PerimeterX's IP-trust scoring every time — don't use them.
PROXY_HOST = os.environ["PROXY_HOST"]
PROXY_PORT = int(os.environ["PROXY_PORT"])
PROXY_USER = os.environ.get("PROXY_USER", "")
PROXY_PASS = os.environ.get("PROXY_PASS", "")


def solve() -> dict:
    # 1) Create the task. Returns immediately with a taskId; the actual
    #    solve runs on Capzy's infrastructure using YOUR proxy.
    task = {
        "type": "AntiPerimeterXTask",
        "websiteURL": "https://example.com",
        "proxyType": "http",
        "proxyAddress": PROXY_HOST,
        "proxyPort": PROXY_PORT,
    }
    if PROXY_USER:
        task["proxyLogin"] = PROXY_USER
        task["proxyPassword"] = PROXY_PASS

    created = requests.post(
        f"{API_BASE}/createTask",
        json={"clientKey": CAPZY_KEY, "task": task},
        timeout=15,
    ).json()

    if created.get("errorId"):
        raise RuntimeError(f"createTask failed: {created.get('errorCode')} — "
                           f"{created.get('errorDescription')}")

    task_id = created["taskId"]
    print(f"created task {task_id}")

    # 2) Poll until ready. Cap the wait at 120s for slower captcha types.
    deadline = time.time() + 120
    while time.time() < deadline:
        result = requests.post(
            f"{API_BASE}/getTaskResult",
            json={"clientKey": CAPZY_KEY, "taskId": task_id},
            timeout=15,
        ).json()

        if result.get("errorId"):
            raise RuntimeError(f"getTaskResult failed: {result.get('errorCode')} — "
                               f"{result.get('errorDescription')}")

        if result["status"] == "ready":
            return result["solution"]

        time.sleep(2)

    raise TimeoutError("solve took longer than 120s")


if __name__ == "__main__":
    solution = solve()
    print("solution:", solution)
    # ─── How to use the result ────────────────────────────────────
    # Set ALL the returned cookies on your HTTP client and reuse the User-Agent. _px3 rotates every ~60 seconds — re-solve when it expires.

File: examples/nodejs/basic.js
/**
 * Solve PerimeterX / HUMAN Security with Capzy — minimal Node.js example.
 *
 * Cost:   from $0.001 per solve (flat)
 * Speed:  ~10 seconds median
 *
 * PerimeterX clearance cookies (_px3 / _px2 / _pxhd) are IP-bound — your
 * proxy is REQUIRED, and there is no ProxyLess variant. Use the SAME
 * sticky proxy your downstream HTTP client will replay the cookies on.
 *
 * Run with (Node 18+):
 *   export CAPZY_KEY="capzy_xxxxxxxxxxxxxxxxxxxxxxxx"
 *   export PROXY_HOST="gw.your-provider.com"
 *   export PROXY_PORT="10000"
 *   export PROXY_USER="your-user"
 *   export PROXY_PASS="your-pass"
 *   node basic.js
 *
 * Uses the built-in global `fetch` — no dependencies, no npm install.
 */

const API_BASE = "https://api.capzy.ai";
const CAPZY_KEY = process.env.CAPZY_KEY;

// Sticky residential / mobile / static-ISP proxy. Datacenter IPs fail
// PerimeterX's IP-trust scoring every time — don't bother.
const PROXY_HOST = process.env.PROXY_HOST;
const PROXY_PORT = parseInt(process.env.PROXY_PORT, 10);
const PROXY_USER = process.env.PROXY_USER || "";
const PROXY_PASS = process.env.PROXY_PASS || "";

async function postJson(path, body) {
  const res = await fetch(`${API_BASE}${path}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  return res.json();
}

async function solve() {
  // 1) Create the task — proxy is required for PerimeterX.
  const task = {
    type: "AntiPerimeterXTask",
    websiteURL: "https://example.com",
    proxyType: "http",
    proxyAddress: PROXY_HOST,
    proxyPort: PROXY_PORT,
  };
  if (PROXY_USER) {
    task.proxyLogin = PROXY_USER;
    task.proxyPassword = PROXY_PASS;
  }
  const created = await postJson("/createTask", {
    clientKey: CAPZY_KEY,
    task,
  });
  if (created.errorId) {
    throw new Error(`createTask: ${created.errorCode} — ${created.errorDescription}`);
  }
  const taskId = created.taskId;
  console.log("created task", taskId);

  // 2) Poll until ready.
  const deadline = Date.now() + 120_000;
  while (Date.now() < deadline) {
    const result = await postJson("/getTaskResult", {
      clientKey: CAPZY_KEY,
      taskId,
    });
    if (result.errorId) {
      throw new Error(`getTaskResult: ${result.errorCode} — ${result.errorDescription}`);
    }
    if (result.status === "ready") return result.solution;
    await new Promise((r) => setTimeout(r, 2000));
  }
  throw new Error("solve took longer than 120s");
}

(async () => {
  const solution = await solve();
  console.log("solution:", solution);
  // ─── Example solution shape ─────────────────────────────────
  // {
  //   token:              "<_px3 cookie value>",
  //   cookie:             "_px3=<value>; _pxhd=<value>; _pxvid=<uuid>",
  //   cookies:            [{ name: "_px3", value: "...", domain: "...", path: "/" }, ...],
  //   userAgent:          "Mozilla/5.0 (...) Chrome/... Safari/...",
  //   challengePresented: true,   // Hold Captcha widget rendered + held
  //   holdDurationSec:    9.4,    // 0 if challengePresented is false
  //   ipBound:            true
  // }
  //
  // ─── How to use the result ──────────────────────────────────
  // Drop-in — paste `cookie` straight into a Cookie: header:
  //
  //   const headers = {
  //     "Cookie":     solution.cookie,
  //     "User-Agent": solution.userAgent,
  //   };
  //   await fetch(targetUrl, { headers, agent: yourProxyAgent });
  //
  // Replay MUST come through the SAME proxy you supplied at solve
  // time (clearance is IP+UA bound). `_px3` rotates every ~60s on
  // newer deployments — re-solve when it expires.
})();
